src/ComposableExpression.jl

```diff
@@ -244,4 +248,264 @@ for op in (
 
+Base.@enum ConstantDerivative::UInt8 Zero One NegOne Other
+
+"""
+    D(ex::AbstractComposableExpression, feature::Integer)
+
+Compute the derivative of `ex` with respect to the `feature`-th variable.
+Returns a new `ComposableExpression` with an expanded set of operators.
+"""
+function D(ex::AbstractComposableExpression, feature::Integer)
+    metadata = DE.get_metadata(ex)
+    raw_metadata = getfield(metadata, :_data)  # TODO: Upstream this so we can load this
+    operators = DE.get_operators(ex)
+    mult_idx = findfirst(==(*), operators.binops)::Integer
+    plus_idx = findfirst(==(+), operators.binops)::Integer
+    nbin = length(operators.binops)
+    nuna = length(operators.unaops)
+    tree = DE.get_contents(ex)
+    operators_with_derivatives = _expand_operators(operators)
+    evaluates_to_constant = map(
+        op -> if op == _zero
+            Zero
+        elseif op == _one
+            One
+        elseif op == _n_one
+            NegOne
+        else
+            Other
+        end, operators_with_derivatives.binops
+    )
+    ctx = SymbolicDerivativeContext(;
+        feature, plus_idx, mult_idx, nbin, nuna, evaluates_to_constant
+    )
+    d_tree = _symbolic_derivative(tree, ctx)
+    return with_metadata(
+        with_contents(ex, d_tree); raw_metadata..., operators=operators_with_derivatives
+    )
+end
+
+Base.@kwdef struct SymbolicDerivativeContext{TUP}
+    feature::Int
+    plus_idx::Int
+    mult_idx::Int
+    nbin::Int
+    nuna::Int
+    evaluates_to_constant::TUP
+end
+
+function _symbolic_derivative(
+    tree::N, ctx::SymbolicDerivativeContext
+) where {T,N<:AbstractExpressionNode{T}}
+    # NOTE: We cannot mutate the tree here! Since we use it twice.
+
+    # Quick test to see if we have any dependence on the feature, so
+    # we can return 0 for the branch
+    any_dependence = any(tree) do node
+        node.degree == 0 && !node.constant && node.feature == ctx.feature
+    end
+
+    if !any_dependence
+        return constructorof(N)(; val=zero(T))
+    elseif tree.degree == 0 # && any_dependence
+        return constructorof(N)(; val=one(T))
+    elseif tree.degree == 1
+        # f(g(x)) => f'(g(x)) * g'(x)
+        f_prime_op = tree.op + ctx.nuna
+
+        ### We do some simplification based on zero/one derivatives ###
+        g_prime = _symbolic_derivative(tree.l, ctx)
+        if g_prime.degree == 0 && g_prime.constant && iszero(g_prime.val)
+            return g_prime
+        else
+            f_prime = constructorof(N)(; op=f_prime_op, l=tree.l)
+
+            if g_prime.degree == 0 && g_prime.constant && isone(g_prime.val)
+                return f_prime
+            else
+                return constructorof(N)(; op=ctx.mult_idx, l=f_prime, r=g_prime)
+            end
+        end
+    else  # tree.degree == 2
+        # f(g(x), h(x)) => f^(1,0)(g(x), h(x)) * g'(x) + f^(0,1)(g(x), h(x)) * h'(x)
+        f_prime_left_op = tree.op + ctx.nbin
+        f_prime_right_op = tree.op + 2 * ctx.nbin
```

Comment:
@Moelf this is where the binary operator indices are shifted. It's nice because it basically lets us generate a symbolic derivative at runtime by literally just adding a couple integers.